src/dump.c

```diff
@@ -1679,77 +1690,75 @@ static jl_value_t *jl_deserialize_value_method(jl_serializer_state *s, jl_value_
         jl_gc_wb(m, m->generator);
     m->invokes = jl_deserialize_value(s, (jl_value_t**)&m->invokes);
     jl_gc_wb(m, m->invokes);
-    m->traced = 0;
     JL_MUTEX_INIT(&m->writelock);
     return (jl_value_t*)m;
 }
 
 static jl_value_t *jl_deserialize_value_method_instance(jl_serializer_state *s, jl_value_t **loc) JL_GC_DISABLED
 {
     int usetable = (s->mode != MODE_IR);
-    jl_method_instance_t *li =
+    jl_method_instance_t *mi =
         (jl_method_instance_t*)jl_gc_alloc(s->ptls, sizeof(jl_method_instance_t),
                                        jl_method_instance_type);
-    memset(li, 0, sizeof(jl_method_instance_t));
+    memset(mi, 0, sizeof(jl_method_instance_t));
     uintptr_t pos = backref_list.len;
     if (usetable)
-        arraylist_push(&backref_list, li);
+        arraylist_push(&backref_list, mi);
     int internal = read_uint8(s->s);
-    int constret = internal & 4;
-    internal &= 0x3;
-    if (internal == 1) {
-        li->min_world = 0;
-        li->max_world = 0;
-    }
-    else if (internal == 2) {
-        li->min_world = jl_world_counter;
-        li->max_world = ~(size_t)0;
-    }
-    else if (internal == 3) {
-        li->min_world = 1;
-        li->max_world = 0;
-    }
-    else if (internal) {
-        assert(0 && "corrupt deserialization state");
-        abort();
-    }
-
-    li->specTypes = (jl_value_t*)jl_deserialize_value(s, (jl_value_t**)&li->specTypes);
-    if (li->specTypes)
-        jl_gc_wb(li, li->specTypes);
-    li->def.value = jl_deserialize_value(s, &li->def.value);
-    if (li->def.value)
-        jl_gc_wb(li, li->def.value);
+    mi->specTypes = (jl_value_t*)jl_deserialize_value(s, (jl_value_t**)&mi->specTypes);
+    jl_gc_wb(mi, mi->specTypes);
+    mi->def.value = jl_deserialize_value(s, &mi->def.value);
+    jl_gc_wb(mi, mi->def.value);
 
     if (!internal) {
         assert(loc != NULL && loc != HT_NOTFOUND);
         arraylist_push(&flagref_list, loc);
         arraylist_push(&flagref_list, (void*)pos);
-        return (jl_value_t*)li;
-    }
-
-    li->inferred = jl_deserialize_value(s, &li->inferred);
-    jl_gc_wb(li, li->inferred);
-    li->inferred_const = jl_deserialize_value(s, &li->inferred_const);
-    if (li->inferred_const)
-        jl_gc_wb(li, li->inferred_const);
-    li->rettype = jl_deserialize_value(s, &li->rettype);
-    jl_gc_wb(li, li->rettype);
-    li->sparam_vals = (jl_svec_t*)jl_deserialize_value(s, (jl_value_t**)&li->sparam_vals);
-    jl_gc_wb(li, li->sparam_vals);
-    li->backedges = (jl_array_t*)jl_deserialize_value(s, (jl_value_t**)&li->backedges);
-    if (li->backedges)
-        jl_gc_wb(li, li->backedges);
-    li->functionObjectsDecls.functionObject = NULL;
-    li->functionObjectsDecls.specFunctionObject = NULL;
-    li->inInference = 0;
-    li->specptr.fptr = NULL;
+        return (jl_value_t*)mi;
+    }
+
+    if (internal == 1) {
+        mi->uninferred = jl_deserialize_value(s, &mi->uninferred);
+        jl_gc_wb(mi, mi->uninferred);
```

Comment:
To me it looks like the order here was (and remains til this day) reversed from serialization: compare https://github.com/JuliaLang/julia/blob/c3235cdacbaabc77d1d5bdb926d2df27f3848af4/src/dump.c#L726-L736 with https://github.com/JuliaLang/julia/blob/c3235cdacbaabc77d1d5bdb926d2df27f3848af4/src/dump.c#L1611-L1627.

If this is indeed a bug, I can submit a fix.